### PR TITLE
[Enhancement] Support metadata query for task

### LINKED ISSUE #6182: Support metadata query for task
We provide an INFORMATION_SCHEMA.tasks to query the Task created by the user, and another INFORMATION_SCHEMA.task_runs to query the execution status of the task run, including history and just created or normally executed task runs.

### PR DESCRIPTION
**Pull request by Astralidea**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #6182

## Problem Summary(Required) ：

We provide an INFORMATION_SCHEMA.tasks to query the task created by the user.
As well as provide the INFORMATION_SCHEMA.task_runs to query the execution status of the task run, including historical and running tasks.

### CODE REVIEW

**Comment by Astralidea** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 22 / 62 (35.48%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/analysis/SchemaTableType.java|0|3|00.00%|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|0|37|00.00%|
|:large_blue_circle:|com/starrocks/catalog/SchemaTable.java|22|22|100.00%|

File: gensrc/thrift/FrontendService.thrift [hunk 1]
```diff
@@ -330,6 +330,40 @@ struct TListTableStatusResult {
...
 }
 
+// Arguments to showTasks/ShowTaskRuns
+struct TShowTasksParams {
+    1: optional string db
+    2: optional Types.TUserIdentity current_user_ident
+}
+
+struct TTaskInfo {
+    1: required string task_name
```
Comment (imay) [inline gensrc/thrift/FrontendService.thrift:340]: should be optional

File: gensrc/thrift/FrontendService.thrift [hunk 2]
```diff
@@ -330,6 +330,40 @@ struct TListTableStatusResult {
...
+    4: optional i64 complete_time
+    5: optional string status
+    6: optional string database_name
+    7: optional string definition
+    8: optional i32 error_code
+    9: optional string error_msg
+}
+
+struct TListTaskRunInfoResult {
+    1: required list<TTaskRunInfo> tables
```
Comment (imay) [inline gensrc/thrift/FrontendService.thrift:364]: all fields should be optional

File: be/src/exec/vectorized/schema_scanner/schema_task_runs_scanner.cpp [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+        //   name,       type,          size,     is_null
+        {"QUERY_ID", TYPE_VARCHAR, sizeof(StringValue), false},
+        {"TASK_NAME", TYPE_VARCHAR, sizeof(StringValue), false},
+        {"CREATE_TIME", TYPE_DATETIME, sizeof(DateTimeValue), true},
+        {"COMPLETE_TIME", TYPE_DATETIME, sizeof(DateTimeValue), true},
+        {"STATUS", TYPE_VARCHAR, sizeof(StringValue), false},
+        {"DATABASE_NAME", TYPE_VARCHAR, sizeof(StringValue), false},
+        {"DEFINITION", TYPE_VARCHAR, sizeof(StringValue), false},
+        {"ERROR_CODE", TYPE_BIGINT, sizeof(StringValue), true},
+        {"ERROR_MSG", TYPE_VARCHAR, sizeof(StringValue), true}};
```
Comment (imay) [inline be/src/exec/vectorized/schema_scanner/schema_task_runs_scanner.cpp:19]: only this is short, why not use ERROR_MESSAGE

File: be/src/exec/vectorized/schema_scanner/schema_task_runs_scanner.cpp [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+namespace starrocks::vectorized {
+
+SchemaScanner::ColumnDesc SchemaTaskRunsScanner::_s_tbls_columns[] = {
+        //   name,       type,          size,     is_null
+        {"QUERY_ID", TYPE_VARCHAR, sizeof(StringValue), false},
+        {"TASK_NAME", TYPE_VARCHAR, sizeof(StringValue), false},
+        {"CREATE_TIME", TYPE_DATETIME, sizeof(DateTimeValue), true},
+        {"COMPLETE_TIME", TYPE_DATETIME, sizeof(DateTimeValue), true},
+        {"STATUS", TYPE_VARCHAR, sizeof(StringValue), false},
+        {"DATABASE_NAME", TYPE_VARCHAR, sizeof(StringValue), false},
```
Comment (imay) [inline be/src/exec/vectorized/schema_scanner/schema_task_runs_scanner.cpp:16]: DATABASE is OK

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 22 / 66 (33.33%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/analysis/SchemaTableType.java|0|3|00.00%|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|0|37|00.00%|
|:large_blue_circle:|com/starrocks/scheduler/TaskRunStatus.java|0|2|00.00%|
|:large_blue_circle:|com/starrocks/scheduler/TaskRunExecutor.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/scheduler/TaskRun.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/catalog/SchemaTable.java|22|22|100.00%|

File: be/src/exec/vectorized/schema_scanner/schema_tasks_scanner.cpp [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+        }
+        default:
+            break;
+        }
+    }
+    _table_index++;
+    return Status::OK();
+}
+
+Status SchemaTasksScanner::get_new_table() {
```
Comment (ABingHuang) [inline be/src/exec/vectorized/schema_scanner/schema_tasks_scanner.cpp:122]: I think get_next_db is a better name?

**Review by ABingHuang — CHANGES_REQUESTED** [review]

File: gensrc/thrift/FrontendService.thrift [hunk 6]
```diff
@@ -330,6 +330,40 @@ struct TListTableStatusResult {
...
+struct TTaskInfo {
+    1: optional string task_name
+    2: optional i64 create_time
+    3: optional string schedule
+    4: optional string database
+    5: optional string definition
+}
+
+struct TListTaskInfoResult {
+    1: required list<TTaskInfo> tables
```
Comment (wyb) [inline gensrc/thrift/FrontendService.thrift:348]:
```suggestion
    1: required list<TTaskInfo> tasks
```

File: gensrc/thrift/FrontendService.thrift [hunk 7]
```diff
@@ -330,6 +330,40 @@ struct TListTableStatusResult {
...
+    4: optional i64 complete_time
+    5: optional string state
+    6: optional string database
+    7: optional string definition
+    8: optional i32 error_code
+    9: optional string error_message
+}
+
+struct TListTaskRunInfoResult {
+    1: optional list<TTaskRunInfo> tables
```
Comment (wyb) [inline gensrc/thrift/FrontendService.thrift:364]:
```suggestion
    1: optional list<TTaskRunInfo> task_runs
```

File: fe/fe-core/src/main/java/com/starrocks/service/FrontendServiceImpl.java [hunk 8]
```diff
@@ -320,6 +328,53 @@ public TListTableStatusResult listTableStatus(TGetTablesParams params) throws TE
...
+    @Override
+    public TListTaskInfoResult showTasks(TShowTasksParams params) throws TException {
+        LOG.debug("get show task request: {}", params);
+        TListTaskInfoResult result = new TListTaskInfoResult();
+        List<TTaskInfo> tasksResult = Lists.newArrayList();
+        result.setTables(tasksResult);
+
+        TaskManager taskManager = GlobalStateMgr.getCurrentState().getTaskManager();
+        List<Task> taskList = taskManager.showTask().stream()
+                .filter(u -> u.getDbName().equals(params.db)).collect(Collectors.toList());
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/service/FrontendServiceImpl.java:340]: Filter db in `taskManager.showTask`？

File: fe/fe-core/src/main/java/com/starrocks/service/FrontendServiceImpl.java [hunk 9]
```diff
@@ -320,6 +328,53 @@ public TListTableStatusResult listTableStatus(TGetTablesParams params) throws TE
         return result;
     }
 
+    @Override
+    public TListTaskInfoResult showTasks(TShowTasksParams params) throws TException {
+        LOG.debug("get show task request: {}", params);
+        TListTaskInfoResult result = new TListTaskInfoResult();
+        List<TTaskInfo> tasksResult = Lists.newArrayList();
+        result.setTables(tasksResult);
+
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/service/FrontendServiceImpl.java:337]: Check privilege.

File: fe/fe-core/src/main/java/com/starrocks/catalog/SchemaTable.java [hunk 10]
```diff
@@ -415,6 +415,35 @@ public static Builder builder() {
...
+                    .put("task_runs",
+                            new SchemaTable(
+                                    SystemIdGenerator.getNextId(),
+                                    "task_runs",
+                                    TableType.SCHEMA,
+                                    builder()
+                                            .column("QUERY_ID", ScalarType.createVarchar(64))
+                                            .column("TASK_NAME", ScalarType.createVarchar(64))
+                                            .column("CREATE_TIME", ScalarType.createType(PrimitiveType.DATETIME))
+                                            .column("COMPLETE_TIME", ScalarType.createType(PrimitiveType.DATETIME))
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/catalog/SchemaTable.java:440]:
```suggestion
                                            .column("FINISH_TIME", ScalarType.createType(PrimitiveType.DATETIME))
```

File: be/src/exec/vectorized/schema_scanner/schema_task_runs_scanner.h [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+    Status get_next(ChunkPtr* chunk, bool* eos) override;
+
+private:
+    Status get_new_table();
+    Status fill_chunk(ChunkPtr* chunk);
+
+    int _db_index{0};
+    int _table_index{0};
+    TGetDbsResult _db_result;
+    TListTaskRunInfoResult _task_result;
```
Comment (wyb) [inline be/src/exec/vectorized/schema_scanner/schema_task_runs_scanner.h:25]:
```suggestion
    TListTaskRunInfoResult _task_run_result;
```

File: be/src/exec/vectorized/schema_scanner/schema_task_runs_scanner.h [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+
+    Status start(RuntimeState* state) override;
+    Status get_next(ChunkPtr* chunk, bool* eos) override;
+
+private:
+    Status get_new_table();
+    Status fill_chunk(ChunkPtr* chunk);
+
+    int _db_index{0};
+    int _table_index{0};
```
Comment (wyb) [inline be/src/exec/vectorized/schema_scanner/schema_task_runs_scanner.h:23]: Status get_new_table();
int _table_index{0};

Please use the correct name.

File: be/src/exec/vectorized/schema_scanner/schema_tasks_scanner.cpp [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+#include "exec/vectorized/schema_scanner/schema_tasks_scanner.h"
+
+#include "exec/vectorized/schema_scanner/schema_helper.h"
+
+namespace starrocks::vectorized {
+
+SchemaScanner::ColumnDesc SchemaTasksScanner::_s_tbls_columns[] = {
+        //   name,       type,          size,     is_null
+        {"TASK_NAME", TYPE_VARCHAR, sizeof(StringValue), true},
+        {"CREATE_TIME", TYPE_DATETIME, sizeof(DateTimeValue), false},
```
Comment (wyb) [inline be/src/exec/vectorized/schema_scanner/schema_tasks_scanner.cpp:12]:
```suggestion
        {"TASK_NAME", TYPE_VARCHAR, sizeof(StringValue), false},
        {"CREATE_TIME", TYPE_DATETIME, sizeof(DateTimeValue), true},
```

File: be/src/exec/vectorized/schema_scanner/schema_task_runs_scanner.cpp [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+                    fill_data_column_with_null(column.get());
+                }
+            }
+            break;
+        }
+        case 9: {
+            // ERROR_MESSAGE
+            {
+                ColumnPtr column = (*chunk)->get_column_by_slot_id(9);
+                const std::string* str = &task_run_info.error_message;
```
Comment (wyb) [inline be/src/exec/vectorized/schema_scanner/schema_task_runs_scanner.cpp:165]: error_message is nullable?

File: fe/fe-core/src/main/java/com/starrocks/service/FrontendServiceImpl.java (hunk 8 above)
Comment (Astralidea) [reply fe/fe-core/src/main/java/com/starrocks/service/FrontendServiceImpl.java:350]: Yes, Task is of DB granularity, and TaskManager is not optimized for DB queries. BE is sent in the order of DB, so FE is filtered here.

File: be/src/exec/vectorized/schema_scanner/schema_task_runs_scanner.cpp (hunk 14 above)
Comment (Astralidea) [reply be/src/exec/vectorized/schema_scanner/schema_task_runs_scanner.cpp:165]: Yes, but there will be no NULL value when FE returns.

**Review by ABingHuang — DISMISSED** [review]
LGTM

**Comment by Astralidea** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 24 / 82 (29.27%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/analysis/SchemaTableType.java|0|2|00.00%|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|0|51|00.00%|
|:large_blue_circle:|com/starrocks/scheduler/TaskRun.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/scheduler/TaskRunStatus.java|1|4|25.00%|
|:large_blue_circle:|com/starrocks/scheduler/TaskRunExecutor.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/catalog/SchemaTable.java|22|22|100.00%|

**Comment by Astralidea** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 39 / 112 (34.82%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/analysis/SchemaTableType.java|0|2|00.00%|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|0|49|00.00%|
|:large_blue_circle:|com/starrocks/scheduler/TaskRun.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/scheduler/TaskRunStatus.java|1|4|25.00%|
|:large_blue_circle:|com/starrocks/scheduler/TaskRunManager.java|4|14|28.57%|
|:large_blue_circle:|com/starrocks/scheduler/TaskRunExecutor.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/scheduler/TaskManager.java|10|17|58.82%|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/catalog/SchemaTable.java|22|22|100.00%|

**Review by wyb — DISMISSED** [review]

**Review by wyb — APPROVED** [review]

**Review by ABingHuang — APPROVED** [review]
LGTM

**Review by chaoyli — APPROVED** [review]

**Pull request merged**